### PR TITLE
Make Ringbuffer not throw StaleSequenceException on read-many operation

### PR DESCRIPTION
**Pull request by jbartok**
The problem we want to fix is that when producers are fast, consumers can have a really hard time keeping- and catching up with them (they often fail to do so, even when they would be able to process fast enough). For example if your head is at 1, and you ask for 100 items from 0, the `Ringbuffer.readManyAsync` call just throws an exception, instead of returning what’s available. By the time the exception travels to the consumer, gets processed and a new request is made, the requested sequence number tends to be stale again and will result in yet another exception. Meanwhile data is being overwritten in the `Ringbuffer` and is lost to the consumer.

The solution aims to remove the `StaleSequenceException` being thrown by the `ReadManyOperation` in such situations and just have it return the data that's available. 

We are able to do this for `Ringbuffer.readManyAsync`, because it returns a `ReadResultSet` carrying sequence numbers, thus allowing the client to notice potential sequence gaps and decide if it can tolerate them or not. On the other had we can't do the same for `Ringbuffer.readOne` because there is no way to observe sequence numbers in that case (it returns just an element from the `Ringbuffer`).

### CODE REVIEW

**Comment by cangencer** [comment]
readManyAsync javadoc would need updating also.

File: hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java [hunk 1]
```diff
@@ -93,14 +92,24 @@ public void next() {
...
+                ReliableTopicMessage message = result.get(i);
 
-                if (cancelled) {
-                    return;
+                long actualSequence = result.getSequence(i);
+                if (actualSequence != sequence) {
+                    if (handleSilentSequenceJump(actualSequence)) {
+                        sequence = actualSequence;
+                    } else {
+                        cancel();
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: How about if we cancel and throw the `StaleSequenceException` back to the user if the listener is not loss tolerant? Now, we just cancel which I guess stops the listener silently without notifying the user.

Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: Can pointed out that this is not possible and we don't throw anything back to the caller even for the current implementation but instead handle the exception in the runner. So ignore this.

**Comment by cangencer** [comment]
verify

**Comment by cangencer** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java [hunk 2]
```diff
@@ -196,34 +199,28 @@ private boolean handleOperationTimeoutException() {
...
         if (listener.isLossTolerant()) {
             if (logger.isFinestEnabled()) {
-                logger.finest("MessageListener " + listener + " on topic: " + topicName + " ran into a stale sequence. "
-                        + "Jumping from oldSequence: " + sequence
-                        + " to sequence: " + headSeq);
+                logger.finest("MessageListener " + listener + " on topic: " + topicName + " ran into a silent sequence"
+                        + " jump from oldSequence: " + sequence + " to sequence: " + newSequence);
             }
-            sequence = headSeq;
+            sequence = newSequence;
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: We will assign this once the method returns so you can remove this.

File: hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java [hunk 3]
```diff
@@ -93,14 +92,24 @@ public void next() {
...
             // we process all messages in batch. So we don't release the thread and reschedule ourselves;
             // but we'll process whatever was received in 1 go.
-            for (Object item : result) {
-                ReliableTopicMessage message = (ReliableTopicMessage) item;
+            for (int i = 0; i < result.size(); i++) {
+                ReliableTopicMessage message = result.get(i);
 
-                if (cancelled) {
-                    return;
+                long actualSequence = result.getSequence(i);
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: Not the topic of this PR but I can see how this `MessageRunner` cannot run with a predicate since it expects a contiguous range of sequences.

File: hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java [hunk 4]
```diff
@@ -53,20 +53,20 @@ public ReadManyOperation(String name, long startSequence, int minSize, int maxSi
 
     @Override
     public void beforeRun() {
-        RingbufferContainer ringbuffer = getRingBufferContainer();
-        ringbuffer.checkBlockableReadSequence(startSequence);
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java]: Do we not check for too-large sequence because we're about to clamp to the largest anyway?

File: hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java [hunk 5]
```diff
@@ -53,20 +53,20 @@ public ReadManyOperation(String name, long startSequence, int minSize, int maxSi
...
+            resultSet = new ReadResultSetImpl<>(minSize, maxSize, getNodeEngine().getSerializationService(), filter);
             sequence = startSequence;
         }
 
         RingbufferContainer ringbuffer = getRingBufferContainer();
+        sequence = ringbuffer.clampReadSequenceToBounds(sequence);
+
         if (minSize == 0) {
-            if (!ringbuffer.shouldWait(sequence)) {
+            if (sequence < ringbuffer.tailSequence() + 1) {
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java]: Can we leave the previous condition? After clamping, we should be fine and it should no longer throw a exception.

File: hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java (hunk 3 above)
Comment (cangencer) [reply hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: We use a different formula in event journal, I think it could be used here too? 
`lostCount = resultSet.getNextSequenceToReadFrom() - resultSet.readCount() - prevSequence`, where prevSequence is the asked for sequence.

Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: Yes, we could report this in the log message below.

File: hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java (hunk 2 above)
Comment (jbartok) [reply hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: Ok, good point, will do.

File: hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java (hunk 3 above)
Comment (jbartok) [reply hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: I'm not 100% sure what you need me to do about it, so pls. check what I'll push.

File: hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java (hunk 4 above)
Comment (jbartok) [reply hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java]: Yes.

File: hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java (hunk 5 above)
Comment (jbartok) [reply hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java]: Ok.

**Comment by jbartok** [comment]
run-lab-run

**Comment by jbartok** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java [hunk 6]
```diff
@@ -101,7 +101,7 @@ public void accept(ReadResultSet<ReliableTopicMessage> result, Throwable throwab
             // but we'll process whatever was received in 1 go.
 
             long lostCount = result.getNextSequenceToReadFrom() - result.readCount() - sequence;
-            if (lostCount > 0 && !isLossTolerable(lostCount)) {
+            if (lostCount != 0 && !isLossTolerable(lostCount)) {
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: how can the loss be minus? when the cluster restarts?

Comment (jbartok) [reply hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: Yes, for example in `ClientReliableTopicOnClusterRestartTest.shouldFail_OnClusterRestart_whenDataLoss_notLossTolerant`. This test was failing before my last commit.

Comment (jbartok) [reply hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: Not sure if we should indeed cancel in this situation though...

**Comment by cangencer** [comment]
@jbartok could you add description of what this fixes?

File: hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java [hunk 7]
```diff
@@ -113,14 +118,11 @@ public void accept(ReadResultSet<ReliableTopicMessage> result, Throwable throwab
                     }
                 }
 
-                sequence++;
             }
+
+            sequence = result.getNextSequenceToReadFrom();
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: you update the member `sequence` only after the iteration is finished but if the exception occurs and it is a terminal exception, the method returns leaving the runner with incorrect sequence number. The master branch code updates the sequence during iteration. Should we revert this back?

File: hazelcast/src/test/java/com/hazelcast/client/ringbuffer/RingbufferTest.java [hunk 8]
```diff
@@ -88,27 +89,62 @@ public void tearDown() {
...
+        assertEquals("10", rs.get(0));
+    }
+
+    @Test
+    public void readManyAsync_whenStartSequenceIsJustBeyondTailSequence() throws Exception {
+        serverRingbuffer.addAllAsync(asList("0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10"), OVERWRITE);
+        CompletionStage<ReadResultSet<String>> f = clientRingbuffer.readManyAsync(11, 1, 10, null);
+
+        //test that the read blocks (should at least fail some of the time, if not the case)
+        TimeUnit.MILLISECONDS.sleep(250);
```
Comment (ihsandemir) [inline hazelcast/src/test/java/com/hazelcast/client/ringbuffer/RingbufferTest.java]: I find the following better understandable:

```
f.toCompletableFuture().get(250, TimeUnit.MILLISECONDS);
```
And add throws to test method `(expected = TimeoutException.class)`

File: hazelcast/src/test/java/com/hazelcast/client/ringbuffer/RingbufferTest.java [hunk 9]
```diff
@@ -88,27 +89,62 @@ public void tearDown() {
...
+        assertFalse(f.toCompletableFuture().isDone());
+    }
+
+    @Test
+    public void readManyAsync_whenStartSequenceIsWellBeyondTailSequence() throws Exception {
+        serverRingbuffer.addAllAsync(asList("0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10"), OVERWRITE);
+        CompletionStage<ReadResultSet<String>> f = clientRingbuffer.readManyAsync(19, 1, 10, null);
+
+        //test that the read blocks (should at least fail some of the time, if not the case)
+        TimeUnit.MILLISECONDS.sleep(250);
```
Comment (ihsandemir) [inline hazelcast/src/test/java/com/hazelcast/client/ringbuffer/RingbufferTest.java]: Same as comment on previous test case.

**Comment by jbartok** [comment]
> @jbartok could you add description of what this fixes?

Added to PR description.

File: hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java [hunk 10]
```diff
@@ -53,18 +53,18 @@ public ReadManyOperation(String name, long startSequence, int minSize, int maxSi
...
-            resultSet = new ReadResultSetImpl<O, O>(minSize, maxSize, getNodeEngine().getSerializationService(), filter);
+            resultSet = new ReadResultSetImpl<>(minSize, maxSize, getNodeEngine().getSerializationService(), filter);
             sequence = startSequence;
         }
 
         RingbufferContainer ringbuffer = getRingBufferContainer();
+        sequence = ringbuffer.clampReadSequenceToBounds(sequence);
+
         if (minSize == 0) {
             if (!ringbuffer.shouldWait(sequence)) {
```
[the remark this answers is not in the document]
Comment (jbartok) [reply hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java]: @mmedenjak commented the opposite at some point, but yes, I think this is more appropriate, except as `sequence < ringbuffer.tailSequence() + 1` (_smaller than_, instead of _not equal_). Will modify.

File: hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java [hunk 11]
```diff
@@ -53,18 +53,18 @@ public ReadManyOperation(String name, long startSequence, int minSize, int maxSi
...
                 readMany(ringbuffer);
```
[the remark this answers is not in the document]
Comment (jbartok) [reply hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java]: Isn't this basically the same thing as your previous comment? Am I missing something?

File: hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java (hunk 7 above)
Comment (jbartok) [reply hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: If the iteration is aborted for some reason, then the runner is also cancelled, so it doesn't matter what the sequence end up like. At least that's how I see it. Am I mistaken?

File: hazelcast/src/test/java/com/hazelcast/client/ringbuffer/RingbufferTest.java (hunk 8 above)
Comment (jbartok) [reply hazelcast/src/test/java/com/hazelcast/client/ringbuffer/RingbufferTest.java]: Ok.

**Comment by mmedenjak** [comment]
@ihsandemir are all review comments addressed?

File: hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java (hunk 7 above)
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: If `terminate(t)` is false, it will not abort, and continue, right?

File: hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java]: No, it is different line. I mean that `readMany` can throw `StaleSequenceException`.

[the remark this answers is not in the document]
Comment (jbartok) [reply hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java]: Ahh, ok. I understand now. We probably can't/shouldn't make `RingbufferContainer.readMany` not throw `StaleSequenceException`, because it's used by the `EventJournal` too, but in `ReadManyOperation.shouldWait`, as far as I can tell, it can actually throw the exception only if the store is not enabled. I would say we should just specify that in the docs. Will do that, let me know if you have a different opinion.

File: hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java (hunk 7 above)
Comment (jbartok) [reply hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: If `terminate(t)` is `false` then it will go on with the `for` loop and once the loop is done it will set the `sequence` correctly. `sequence` is no longer used anywhere inside the loop, so this should be fine, right?

File: hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (jbartok) [reply hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java]: Actually no, I'm mistaken, it won't throw the exception even then (when the store is enabled). Yes clamping is not happening in that case, but the sequence won't be considered stale, the exception won't be thrown, it will instead attempt to load it from the store (and just return `null` if it can't).

[the remark this answers is not in the document]
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java]: Even if you clamped the sequence number, until you do the `RingbufferContainer.readMany` call, is it not possible that another thread did a number of readMany's and your clamped sequence number became stale again?

File: hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java (hunk 3 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: Looks good :)

File: hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java (hunk 6 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: I'm ok with failing this case as well. Ideally, we would like to fail readers even if the new ringbuffer gets filled enough to catch up with the requested sequence see [here](https://github.com/hazelcast/hazelcast/issues/11857#issuecomment-352401655). 
This here is a best-effort approach for now.

File: hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java]: Between `ringbuffer.shouldWait(sequence)` and `sequence != ringbuffer.tailSequence() + 1`, I don't mind which one we use. So I'm ok with the above change. But, let's try to analyse if `ringbuffer.shouldWait(sequence)` can, in fact, throw an exception.

In case the ringbuffer store is not configured, `ringbuffer.clampReadSequenceToBounds(sequence)` will change the `sequence` to a valid sequence in the range `[headSequence, tailSequence + 1]`. In the case when the ringbuffer store is enabled, it will not fix any sequences below `headSequence` because it assumes that the store has older sequences.

After clamping, we invoke `ringbuffer.shouldWait(sequence)`. The only method here which may throw an exception is `RingbufferContainer#checkBlockableReadSequence` and in two cases:
• `isTooLargeSequence(readSequence)` returns true -> this means `sequence > tailSequence() + 1` but this is not possible because we cover this case in `clampReadSequenceToBounds` and there is no concurrent mutation on the ringbuffer since it's all done in the partition thread
• `isStaleSequence(readSequence)` returns true -> this means `sequence < headSequence() && !store.isEnabled()` returns true. This too is not possible because it too is covered by the clamping: either the sequence is equal or larger than the head sequence or the store is enabled - such is the contract of the `clampReadSequenceToBounds` method. 

In summary, I think `ringbuffer.shouldWait(sequence)` can throw an exception after `ringbuffer.clampReadSequenceToBounds(sequence)` is invoked but I don't mind using 
`sequence < ringbuffer.tailSequence() + 1` instead.

File: hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java]: It is correct that we want to avoid the `StaleSequenceException` in this PR but we want to do this by clamping and relying that we pass any subsequent checks, instead of removing those checks.

With that in mind, I don't see a problem here, unless there is some concurrent mutation on the ringbuffer, same as in the previous case - https://github.com/hazelcast/hazelcast/pull/16303#discussion_r363743868 and I agree with @jbartok . 

AFAIK the entire `ReadManyOperation#shouldWait` is executed on the partition thread in it's entirety and items can be added to the same ringbuffer only once the partition thread is freed.

So, if clamping will restrict the sequence to [headSequence, tailSequence] (or less than headSequence if store is enabled), since there cannot be any concurrent mutation on the ringbuffer, since no other thread than the partition thread can invoke `readMany` and since `checkBlockableReadSequence`  throws an exception only if `sequence > tailSequence() + 1` (which is not possible because of clamping) or if `sequence < headSequence() && !store.isEnabled()` (which again is not possible because of clamping), I don't see how an exception can be thrown here.

It is possible, though, that the `shouldWait()` method may return (because there are not enough items) after which new items might be added, and the `shouldWait()` is executed again. This, again, is not an issue since we clamp every time we execute the method.

@ihsandemir can you elaborate with a specific scenario and thread scheduling how we might encounter an exception here?

File: hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java (hunk 7 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java]: Seems fine to me.

File: hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java]: Ok, I understand now. I was assuming that concurrent mutation was possible but since it runs on the certain partition thread, I think that will not be a problem.

**Comment by mmedenjak** [comment]
Thank you for the PR and the reviews, everyone!
